Normalize pixel values (0-255 → 0-1)
Reshape data for CNN input (28x28x1)
Split into train/validation/test sets
Create data augmentation
Save preprocessed data


In [1]:
print("hello world")

hello world


In [2]:
#imports 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

In [3]:
# tensorflow import
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [4]:

#random seed for reporoductivity (default 42)
np.random.seed(42)

#loading csv files 
train_data = pd.read_csv('../data/sign_mnist_train.csv')
test_data = pd.read_csv('../data/sign_mnist_test.csv')

#separating into feats and lables 
X_train_raw = train_data.drop('label', axis=1).values
y_train_raw = train_data['label'].values

X_test = test_data.drop('label', axis=1).values
y_test = test_data['label'].values

In [5]:
#print data before normalizing for comparisonn
print(f"\nBefore normalization:")
print(f"Min pixel value: {X_train_raw.min()}")
print(f"Max pixel value: {X_train_raw.max()}")
print(f"Mean pixel value: {X_train_raw.mean():.2f}")

#normalize to 0-1 
X_train_normalized = X_train_raw / 255.0
X_test_normalized = X_test / 255.0

#pritn after normalization for comparison 
print(f"\nAfter normalization:")
print(f"Min pixel value: {X_train_normalized.min()}")
print(f"Max pixel value: {X_train_normalized.max()}")
print(f"Mean pixel value: {X_train_normalized.mean():.3f}")


Before normalization:
Min pixel value: 0
Max pixel value: 255
Mean pixel value: 159.29

After normalization:
Min pixel value: 0.0
Max pixel value: 1.0
Mean pixel value: 0.625


### results

Before normalization:
Min pixel value: 0
Max pixel value: 255
Mean pixel value: 159.29

After normalization:
Min pixel value: 0.0
Max pixel value: 1.0
Mean pixel value: 0.625

### evaluation 

In [ ]:
# reshape data for cnn 

#first we'll print data before reshaping 
print(f"X_train shape: {X_train_normalized.shape} (flat: 784 pixels)")

# Reshape from (N, 784) to (N, 28, 28, 1)
# 28x28 = image dimensions
# 1 = grayscale (single channel)
X_train_reshaped = X_train_normalized.reshape(-1, 28, 28, 1)
X_test_reshaped = X_test_normalized.reshape(-1, 28, 28, 1)

print(f"\nAfter reshaping:")
print(f"X_train shape: {X_train_reshaped.shape} (28x28x1 images)")
print(f"X_test shape: {X_test_reshaped.shape}")
